### PR TITLE
introduce terraform azdo workflow

### PR DESCRIPTION
**Pull request by HadwaAbdelhalem**
This PR addresses #658 by:

- Updates azdo provider to set the needed terraform remote state variables as part of azd pipeline config --provider azdo cmd.
- Introduces an additional azdo pipeline for terraform, and update the existing azdo pipeline to remove the terraform related variables.
- Updates all templates todo/repoman.yaml files to use the right azdo pipeline based on the infra selection.

image

Co-authored-by: Hattan Shobokshi <hattan.shobokshi@microsoft.com>

### CODE REVIEW

File: cli/azd/pkg/azdo/pipeline.go [hunk 1]
```diff
@@ -152,8 +158,20 @@ func getDefinitionVariables(
 		variables["ARM_TENANT_ID"] = createBuildDefinitionVariable(credentials.TenantId, false, false)
 		variables["ARM_CLIENT_ID"] = createBuildDefinitionVariable(credentials.ClientId, true, false)
 		variables["ARM_CLIENT_SECRET"] = createBuildDefinitionVariable(credentials.ClientSecret, true, false)
+
+		// Sets the terraform remote state environment variables in github
```
Comment (vhvb1989) [inline cli/azd/pkg/azdo/pipeline.go:162]: not github, right?

File: cli/azd/pkg/azdo/pipeline.go [hunk 2]
```diff
@@ -140,7 +146,7 @@ func CreatePipeline(
 func getDefinitionVariables(
 	env *environment.Environment,
 	credentials AzureServicePrincipalCredentials,
-	provisioningProvider provisioning.Options) *map[string]build.BuildDefinitionVariable {
+	provisioningProvider provisioning.Options) (*map[string]build.BuildDefinitionVariable, error) {
```
Comment (vhvb1989) [inline cli/azd/pkg/azdo/pipeline.go:149]: Do we also need to check if `provider.conf.json` is in the infra folder?
Can we get the values we need from there?  (if they are not in the env? )

Comment (HadwaAbdelhalem) [reply cli/azd/pkg/azdo/pipeline.go:149]: @vhvb1989 we do not store the values directly in the provider.conf.json. we set the values in the Json to variables that gets populated from env vars instead to allow the backend config values to change based on the selected azd env . see the provider.conf.json  sample [here](https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/use-terraform-for-azd)

File: cli/azd/pkg/azdo/pipeline.go (hunk 1 above)
Comment (HadwaAbdelhalem) [reply cli/azd/pkg/azdo/pipeline.go]: fixed

File: templates/todo/projects/python-mongo/.repo/terraform/repo.yaml [hunk 3]
```diff
@@ -109,6 +109,14 @@ repo:
         - .github/**/*
         - .devcontainer/**/*
         - "infra/**/*"
+        - .azdo/**/*
```
Comment (vhvb1989) [inline templates/todo/projects/python-mongo/.repo/terraform/repo.yaml:112]: Can you ignore only `.azdo/pipelines/*/azure-dev.yml` ?   So we can still have the README.md copied for all?  
Or is there a reason why we want to exclude everything from .azdo/ folder?

Comment (HadwaAbdelhalem) [reply templates/todo/projects/python-mongo/.repo/terraform/repo.yaml]: good find @vhvb1989, updated it

**Review by vhvb1989 — APPROVED** [review]
Thank you very much

**Pull request merged**